### ISSUE TITLE
Dataloader behavior changed in 0.4.1

### ISSUE THREAD

**Issue by zcqian** [pytorch/pytorch#10979]
## Issue description

Given the same random seed, `DataLoader` shuffles differently in version 0.4.1 (and in git HEAD). It was consistent across versions 0.3.0, 0.3.1, and 0.4.0.

If this is the intended behavior, please close the issue, thanks.

## Code example
```py
import torch
from torch.utils.data import DataLoader, TensorDataset

torch.manual_seed(10001)

X = torch.rand(100)
y = torch.LongTensor(100).random_(0, 10)

dataloader = DataLoader(TensorDataset(X, y), batch_size=5, shuffle=True)

for batch_idx, (data, target) in enumerate(dataloader):
    print("Batch: %2d" % batch_idx, "Target:", target.numpy())

print("PyTorch:", torch.__version__)
```
## System Info
- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): conda
- OS: macOS 10.14
- PyTorch version: 0.4.1
- Python version: Python 3.6.6 |Anaconda, Inc.| (default, Jun 28 2018, 11:07:29)

**Comment by SsnL** [comment]
It is possible that this got changed. We try to maintain reproducible random number generation, but it is not guaranteed as it would make optimizing existing kernels impossible. So I think it is good to close this. Feel free to reopen if you find anything related that looks wrong.
